extensions-core/spark-extensions/src/main/scala/org/apache/druid/spark/clients/DruidClient.scala

```diff
@@ -0,0 +1,180 @@
+  mapAsScalaMapConverter, seqAsJavaListConverter}
+
+/**
+  * Going with SegmentMetadataQueries despite the significant overhead because there's no other way
+  * to get accurate information about Druid columns.
+  *
+  * This is pulled pretty directly from the druid-hadoop-input-format project on GitHub. There is
+  * likely substantial room for improvement.
+  *
+  * @param httpClient
+  * @param objectMapper
+  * @param hostAndPort
+  */
+class DruidClient(
+                   val httpClient: HttpClient,
+                   val objectMapper: ObjectMapper,
+                   val hostAndPort: HostAndPort
+                 ) extends Logging {
+  private val RETRY_COUNT = 5
+  private val RETRY_WAIT_SECONDS = 5
+  private val TIMEOUT_MILLISECONDS = 300000
+  private val druidBaseQueryURL: HostAndPort => String =
+    (hostAndPort: HostAndPort) => s"http://$hostAndPort/druid/v2/"
+
+  /**
+    * The SQL system catalog tables are incorrect for multivalue columns and don't have accurate
+    * type info, so we need to fall back to segmentMetadataQueries. Given a DATASOURCE and a range
+    * of INTERVALS to query over, return a map from column name to a tuple of
+    * (columnType, hasMultipleValues). Note that this is a very expensive operation over large
+    * numbers of segments. If possible, this method should only be called with the most granular
+    * starting and ending intervals instead of over a larger interval.
+    *
+    * @param dataSource The Druid dataSource to fetch the schema for.
+    * @param intervals  The intervals to return the schema for.
+    * @return A map from column name to data type and whether or not the column is multi-value
+    *         for the schema of DATASOURCE.
+    */
+  def getSchema(dataSource: String, intervals: List[Interval]): Map[String, (String, Boolean)] = {
+    val body = Druids.newSegmentMetadataQueryBuilder()
+      .dataSource(dataSource)
+      .intervals(intervals.asJava)
+      .analysisTypes(SegmentMetadataQuery.AnalysisType.SIZE)
+      .merge(true)
+      .context(Map[String, AnyRef](
+        "timeout" -> Int.box(TIMEOUT_MILLISECONDS)
+      ).asJava)
+      .build()
+    val response = sendRequestWithRetry(
+      druidBaseQueryURL(hostAndPort), RETRY_COUNT, Option(objectMapper.writeValueAsBytes(body))
+    )
+    val segments =
+      objectMapper.readValue[JList[SegmentAnalysis]](
+        response.getContent, new TypeReference[JList[SegmentAnalysis]]() {})
+    if (segments.size() == 0) {
+      throw new ISE(
+        s"No segments found for intervals [${intervals.mkString(",")}] on $dataSource"
+      )
+    }
+    // Since we're setting merge to true, there should only be one item in the list
+    if (segments.size() > 1) {
+      logWarn("Segment metadata response had more than one row, problem?")
+    }
+    segments.asScala.head.getColumns.asScala.map{ case (name: String, col: ColumnAnalysis) =>
```

Comment:
I couldn't find it in the docs, but what happens if segments have different schemas. For ex - a dimension maybe a string in one segment and a double in another. Should we set `merge` to `false` and give precedence to the latest segment when determining a column's type and other information.

Reply:
[The analysis will error if a dimension has different types across two merged segments](https://github.com/apache/druid/blob/master/processing/src/main/java/org/apache/druid/query/metadata/metadata/ColumnAnalysis.java#L138). The [returned column will have type `String` ](https://github.com/apache/druid/blob/master/processing/src/main/java/org/apache/druid/query/metadata/metadata/ColumnAnalysis.java#L36) and so we'll silently widen the type to String. It's been a couple years since I wrote this piece, and yeah that is really non-obvious (I had to go code spelunking to figure it out). I'm inclined to add a comment to the code explaining what happens and then leave the current behavior, but I could also see just throwing an error back to the caller and refusing to continue.

Reply:
To the second part of your question, the reason we don't just take the last segment in an interval and use its schema is because the latest segment may happen to not have multiple values for a particular column where other segments do, or a column may be all null for the latest segment and so not show up in the Segment Metadata. We query the full range of the requested interval and then merge to handle these cases, which does open up the merging problem if types change. At least in my Druid experience, the first two cases are much more frequent than the third case.